pkg/controller/certificatesigningrequest/approver_controller.go

```diff
@@ -0,0 +1,335 @@
+		return nil
+	}
+	for _, a := range c.approvers {
+		if a.recognize(csr) {
+			return a.verify(csr)
+		}
+	}
+	klog.Warningf("unrecognized CertificateSigningRequest %s", key)
+	return nil
+}
+
+func appendApprovalCondition(csr *certificatesv1.CertificateSigningRequest, message string) {
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:    certificatesv1.CertificateApproved,
+		Status:  corev1.ConditionTrue,
+		Reason:  "AutoApproved",
+		Message: message,
+	})
+}
+
+func appendDeniedCondition(csr *certificatesv1.CertificateSigningRequest, message string) {
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:    certificatesv1.CertificateDenied,
+		Status:  corev1.ConditionTrue,
+		Reason:  "ValidationFailed",
+		Message: message,
+	})
+}
+
+func decodeCertificateRequest(pemBytes []byte) (*x509.CertificateRequest, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil || block.Type != certutil.CertificateRequestBlockType {
+		err := fmt.Errorf("PEM block type must be %s", certutil.CertificateRequestBlockType)
+		return nil, err
+	}
+	return x509.ParseCertificateRequest(block.Bytes)
+}
+
+func (c *CSRApproverController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.syncCSR(key.(string))
+	if err != nil {
+		// Put the item back in the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Failed to sync CertificateSigningRequest", "CertificateSigningRequest", key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+type ipsecCertificateApprover struct {
+	client        clientset.Interface
+	approvePolicy config.AutoApprovePolicy
+}
+
+var _ approver = (*ipsecCertificateApprover)(nil)
+
+func (ic *ipsecCertificateApprover) recognize(csr *certificatesv1.CertificateSigningRequest) bool {
+	if csr.Spec.SignerName != antreaSignerName {
+		return false
+	}
+	for _, usage := range csr.Spec.Usages {
+		if !ipsecTunnelUsages.Has(string(usage)) {
+			return false
+		}
+	}
+	return true
+}
+
+func (ic *ipsecCertificateApprover) verify(csr *certificatesv1.CertificateSigningRequest) error {
+	// Do not approve or deny the requests if AutoApprovePolicy is set to Never.
+	// User should approve or deny the requests manually to get the CSR signed.
+	if ic.approvePolicy == config.AutoApprovePolicyNever {
+		return nil
+	}
+	var deniedReasons []string
+	cr, err := decodeCertificateRequest(csr.Spec.Request)
+	if err != nil {
+		return err
+	}
+	if err := verifyIPSecCSR(ic.client, cr, csr.Spec.Usages); err != nil {
+		if _, ok := err.(*transientError); ok {
+			return err
+		} else {
+			deniedReasons = append(deniedReasons, err.Error())
+		}
+	}
+	if err := verifyPodOnNode(ic.client, cr.Subject.CommonName, csr); err != nil {
```

Comment:
should be more straighforward to make `verifyIPSecCSR` and `verifyPodOnNode` member methods of `ipsecCertificateApprover` as they rely on the member variables.